aggregator/src/aggregator/aggregation_job_writer.rs

```diff
@@ -332,15 +333,53 @@ impl<const SEED_SIZE: usize, Q: CollectableQueryType, A: vdaf::Aggregator<SEED_S
                 for (aggregation_job_id, report_aggregation_ords) in by_aggregation_job_index.iter()
                 {
                     // unwrap safety: index lookup
-                    let (op, agg_job, report_aggs) =
+                    let (agg_job_op, agg_job, report_aggs) =
                         by_aggregation_job.get(aggregation_job_id).unwrap();
-                    if op == &Operation::Put
+                    if agg_job_op == &Operation::Put
                         && matches!(agg_job.state(), AggregationJobState::InProgress)
                     {
                         outstanding_aggregation_jobs += 1;
-                    } else if op == &Operation::Update
+                    } else if agg_job_op == &Operation::Update
                         && !matches!(agg_job.state(), AggregationJobState::InProgress)
                     {
+                        // GC hack: if we are Putting the batch, that means that it does not exist
+                        // in the datastore. But since we first write the batch when an aggregation
+                        // job referencing that batch is created, and we are completing the
+                        // aggregation job here, we must have deleted the batch at some point
+                        // between creating & completing this aggregation job. This should only be
+                        // possible if the batch is GC'ed, as part of a time-interval task. In that
+                        // case, it is acceptable to skip writing the batch entirely; and indeed, we
+                        // must do so, since otherwise we might underflow the
+                        // outstanding_aggregation_jobs counter.
+                        //
+                        // See https://github.com/divviup/janus/issues/2464 for more detail.
+                        if batch_op == Operation::Put {
+                            // Guard to ensure we are in the situation we think we're in:
+                            // `to_batch_interval` returns a value only for the time-interval query
+                            // type; this code effectively checks that the query type is
+                            // time-interval, and the batch interval is past the GC window.
+                            if !Q::to_batch_interval(batch_identifier)
+                                .map(|interval| interval.end() < tx.clock().now())
+                                .unwrap_or(false)
```

Comment:
This works fine, but we have `self.task` in scope, so can't we more directly check the query type?

Reply:
I suppose we should check both -- I also want to guard against a time-interval task unexpectedly not finding a batch that shouldn't be GC'ed.